File: src/dtrace2systemtap.pl
#!/usr/bin/perl
#
# Copyright (C) 2011 Red Hat, Inc.
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
#
# Author: Daniel P. Berrange <berrange@redhat.com>
#
# Generate a set of systemtap probe definitions corresponding to
# DTrace probe markers in libvirt.so
#
#  perl dtrace2systemtap.pl probes.d > libvirt_probes.stp
#

use strict;
use warnings;

my $file;
my @files;
my %files;

my $bindir = shift @ARGV;
my $sbindir = shift @ARGV;
my $libdir = shift @ARGV;

my $probe;
my $args;

# Read the DTraceprobes definition
while (<>) {
    next if m,^\s*$,;

    next if /^\s*provider\s+\w+\s*{\s*$/;
    next if /^\s*};\s*$/;

    if (m,^\s*\#,) {
	if (m,^\s*\#\s*file:\s*(\S+)\s*$,) {
	    $file = $1;
	    push @files, $file;
	    $files{$file} = { prefix => undef, probes => [] };
	} elsif (m,^\s*\#\s*prefix:\s*(\S+)\s*$,) {
	    $files{$file}->{prefix} = $1;
	} elsif (m,^\s*\#\s*binary:\s*(\S+)\s*$,) {
	    $files{$file}->{binary} = $1;
	} else {
	    # ignore unknown comments
	}
    } else {
	if (m,\s*probe\s+([a-zA-Z0-9_]+)\((.*?)(\);)?$,) {
	    $probe = $1;
	    $args = $2;
	    if ($3) {
		push @{$files{$file}->{probes}}, [$probe, $args];
		$probe = $args = undef;
	    }
	} elsif ($probe) {
	    if (m,^(.*?)(\);)?$,) {
		$args .= $1;
		if ($2) {
		    push @{$files{$file}->{probes}}, [$probe, $args];
		    $probe = $args = undef;
		}
	    } else {
		die "unexpected data $_ on line $.";
	    }
	} else {
	    die "unexpected data $_ on line $.";
	}
    }
}

# Write out the SystemTap probes
foreach my $file (@files) {
    my $prefix = $files{$file}->{prefix};
    my @probes = @{$files{$file}->{probes}};

    print "# $file\n\n";
    foreach my $probe (@probes) {
	my $name = $probe->[0];
	my $args = $probe->[1];

	my $pname = $name;
	$pname =~ s/${prefix}_/libvirt.$prefix./;

	my $binary = "$libdir/libvirt.so";
	if (exists $files{$file}->{binary}) {
	    $binary = $sbindir . "/" . $files{$file}->{binary};
	}

	print "probe $pname = process(\"$binary\").mark(\"$name\") {\n";

	my @args = split /,/, $args;
	for (my $i = 0 ; $i <= $#args ; $i++) {
	    my $arg = $args[$i];
	    my $isstr = $arg =~ /char\s+\*/;
	    $arg =~ s/^.*\s\*?(\S+)$/$1/;

	    if ($isstr) {
		print "  $arg = user_string(\$arg", $i + 1, ");\n";
	    } else {
		print "  $arg = \$arg", $i + 1, ";\n";
	    }
	}
	print "}\n\n";
    }
    print "\n";
}

File: scripts/dtrace2systemtap.py
#!/usr/bin/env python3
#
# Copyright (C) 2011-2019 Red Hat, Inc.
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library.  If not, see
# <http://www.gnu.org/licenses/>.
#
#
# Generate a set of systemtap probe definitions corresponding to
# DTrace probe markers in libvirt.so
#
#  python dtrace2systemtap.py probes.d > libvirt_probes.stp
#

import re
import sys

file = None
filelist = []
files = {}

bindir = sys.argv[1]
sbindir = sys.argv[2]
libdir = sys.argv[3]
dtrace = sys.argv[4]

probe = None
args = None

# Read the DTraceprobes definition

with open(dtrace, "r") as fh:
    lineno = 0
    for line in fh:
        lineno = lineno + 1
        line = line.strip()
        if line == "":
            continue
        if "provider " in line and "{" in line:
            continue
        if "};" in line:
            continue

        if line.startswith("#"):
            m = re.search(r'''^\#\s*file:\s*(\S+)$''', line)
            if m is not None:
                file = m.group(1)
                filelist.append(file)
                files[file] = {"prefix": None, "probes": []}
                continue

            m = re.search(r'''^\#\s*prefix:\s*(\S+)$''', line)
            if m is not None:
                files[file]["prefix"] = m.group(1)
                continue

            m = re.search(r'''^\#\s*binary:\s*(\S+)$''', line)
            if m is not None:
                files[file]["binary"] = m.group(1)
                continue

            m = re.search(r'''^\#\s*module:\s*(\S+)$''', line)
            if m is not None:
                files[file]["module"] = m.group(1)

            # ignore unknown comments
        else:
            m = re.search(r'''probe\s+([a-zA-Z0-9_]+)\((.*?)(\);)?$''', line)
            if m is not None:
                probe = m.group(1)
                args = m.group(2)
                if m.group(3) is not None:
                    files[file]["probes"].append([probe, args])
                    probe = None
                    args = None
            elif probe is not None:
                m = re.search(r'''^(.*?)(\);)?$''', line)
                if m is not None:
                    args = args + m.group(1)
                    if m.group(2) is not None:
                        files[file]["probes"].append([probe, args])
                        probe = None
                        args = None
                else:
                    raise Exception("unexpected data %s on line %d" %
                                    (line, lineno))
            else:
                raise Exception("unexpected data %s on line %d" %
                                (line, lineno))

# Write out the SystemTap probes
for file in filelist:
    prefix = files[file]["prefix"]
    probes = files[file]["probes"]

    print("# %s\n" % file)
    for probe in probes:
        name = probe[0]
        args = probe[1]

        pname = name.replace(prefix + "_", "libvirt." + prefix + ".")

        binary = libdir + "/libvirt.so"
        if "binary" in files[file]:
            binary = sbindir + "/" + files[file]["binary"]
        if "module" in files[file]:
            binary = libdir + "/" + files[file]["module"]

        print("probe %s = process(\"%s\").mark(\"%s\") {" %
              (pname, binary, name))

        argbits = args.split(",")
        for idx in range(len(argbits)):
            arg = argbits[idx]
            isstr = False
            if 'char *' in arg:
                isstr = True

            m = re.search(r'''^.*\s\*?(\S+)$''', arg)
            if m is not None:
                arg = m.group(1)
            else:
                raise Exception("Malformed arg %s" % arg)

            if isstr:
                print("  %s = user_string($arg%d);" % (arg, idx + 1))
            else:
                print("  %s = $arg%d;" % (arg, idx + 1))
        print("}\n")
    print("")
